src/ucs/memory/memtype_cache.c

```diff
@@ -127,25 +134,59 @@ UCS_PROFILE_FUNC_VOID(ucs_memtype_cache_update_internal,
     }
 
     if (action == UCS_MEMTYPE_CACHE_ACTION_SET_MEMTYPE) {
-        ucs_memtype_cache_insert(memtype_cache, start, end, mem_type);
-    }
-
-    /* slice old regions by the new region, to preserve the previous memory type
-     * of the non-overlapping parts
-     */
-    ucs_list_for_each_safe(region, tmp, &region_list, list) {
-        if (start > region->super.start) {
-            /* create previous region */
-            ucs_memtype_cache_insert(memtype_cache, region->super.start, start,
-                                     region->mem_type);
-        }
-        if (end < region->super.end) {
-            /* create next region */
-            ucs_memtype_cache_insert(memtype_cache, end, region->super.end,
-                                     region->mem_type);
+        /* merge found regions into signle one (if any):
+         * - calculate start/end of the region constructed
+         *   after merging all regions which intersect with
+         *   new one
+         * - free old region objects
+         * - create and insert the new region */
+
+        /* we have to save real buffer start/end addresses */
+        start = (uintptr_t)address;
+        end   = start + size;
+
+        ucs_list_for_each_safe(region, tmp, &region_list, list) {
+            /* find smallest start address */
+            start = ((region->buf_start < start) ?
+                     region->buf_start : start);
+
+            /* find biggest end address */
+            end = ((region->buf_end > end) ?
+                   region->buf_end : end);
+
```

Comment:
can we add assert (mem_type == region->mem_type) ?

Reply:
what happens if two regions with shared page with different mem types.  we still have the issue right ?because we can't merge them

Reply:
> what happens if two regions with shared page with different mem types

is it a possible situation? can we have a situation that both `cudaMalloc()`/`cudaMallocManaged()` and `malloc()` returns memory where a one page is used to reflect host memory and GPU device memory?

> we still have the issue right ?because we can't merge them

if an answer on my question above - yes, then yes we can't merge then and we should handle it somehow

Reply:
I think it may be possible scenario if you use cuda API to allocate different type of memory (cudaMalloc, cudaMalloManaged, cudaMallocHost).

is it possible to keep the both region in cache w/o slicing  the last page of 1st region when 2nd regions with fist page is shared is inserted.  The search would work in scenario right?

Reply:
> it possible to keep the both region in cache w/o slicing the last page of 1st region when 2nd regions with fist page is shared is inserted. The search would work in scenario right?

I'll check it

Reply:
@bureddy i'm going to simulate this situation in UCS MemType cache gtest by generating two fake `ALLOC` events with different memory types (e.g. `HOST` and `CUDA`) that will share one page. This will shed a light on how the current implementation of UCS MemType cache and UCS Page table can handle such situations